### DISCUSSION TITLE
Extracting accounting information from a container

### DISCUSSION THREAD

**Discussion by yaqwsx** [containers/podman#10156] [Q&A]
Hello,
I would like to be able to extract accounting information from a running or stopped rootless container (e.g., memory usage, CPU time/cycles used). The `podman inspect` does not give any of this information; `podman stats` gives only information about memory (in human-readable units) but no information about consumed CPU cycles. Is there any other podman command that can give me this information?

Since I wasn't able to get the accounting information from podman commands, I tried  to inspect the group mounted in the container under `/sys/fs/cgroup`. When I examine these files from the container, they give me the information I want. However, I am struggling with an effective way of reading them outside container. The `podman cp` reports files under `/sys` as missing. I could in theory use `podman exec <cId> cat /sys/fs/cgroup/cpu.stats` but that affects the statistics (probably negligible, but still). What is a nice and clean way to read out the accounting information from groups?

**Comment by mheon** [comment]
The `podman top` command seems like it would do what you want - both memory and CPU cycles are available as format descriptors.

**Reply by mheon** [reply]
(You would  need to add results for all processes in the container together to get an actual cumulative figure for the whole container, though...)

**Reply by yaqwsx** [reply]
Thank you for your answer. However, `podman top` does not do what I exactly want. It gives you utilization and wall-clock time (if I don't miss something). What I would like to get is user, system, and wall-clock time and maximum memory usage from a command run inside the container (basically what the utility time does).

I am experimenting with using Podman as a reproducible specification of benchmarking task - this is why I am interested in accounting. There are already tools using groups for this (e.g., https://github.com/sosy-lab/benchexec), but they struggle from the absence of easy reproducibility of the execution environment

The groups basically do all the accounting for me. They also do it precisely, however, I am not sure what is the proper way to extract them. Especially when the container stops.

**Reply by mheon** [reply]
Alright. The information you require should be available directly in the host's cgroups, without requiring `podman exec` to get into the container. Exactly where it exists depends on the configuration on your system - on my cgroupsv2 + systemd system, for example, root containers have their cgroups in `/sys/fs/cgroup/machine.slice/libpod-<CONTAINERID>/`.

**Reply by yaqwsx** [reply]
I found out that in rootless mode this location is usually `/sys/fs/cgroup/user.slice/user-$ID.slice/user@$ID.service/`. But I guess I could create a fresh cgroup and specify it via `--cgroup parent`. Then I could just watch for the accounting information of the parent group. I will give it a try.

What I might see as a possible pitfall is that podman creates a subgroup `libpod-<hash>` and `libpod-<hash>/container` inside (where are the tasks from the container). I am not sure what is their purpose - does podman run something other than the processes in the `libpod-<hash>`group? Can I rely on the fact that the accounting information for the specified parent will match the information inside `libpod-<hash>/container`?

**Reply by mheon** [reply]
The `container` subgroup is an implementation detail specific to some OCI runtimes. It shouldn't matter, though - the parent cgroup should have cumulative statistics that include all sub-cgroups (at least on cgroups v2, where cgroups were redesigned to be sane).

**Reply by yaqwsx** [reply]
Perfect, thanks for the confirmation!

**Reply by rhatdan** [reply]
This might be something we would think about extending podman stats to gather and present to the user.

**Reply by rhatdan** [reply]
podman stats --help
Display a live stream of container resource usage statistics

Description:
  Display percentage of CPU, memory, network I/O, block I/O and PIDs for one or more containers.

**Reply by yaqwsx** [reply]
Thank you for pointing me in the right direction. I think the procedure mentioned above will work; I have just stumbled upon this (probably a bug): #10173 The workaround would be to execute podman from a process in the target cgroup, however, that would affect the collected statistics.

@rhatdan I think that would be useful for my use case; however, I am not sure if there is a widespread demand for such a feature.
